EIPS/draft-eip-multi-algorithm-pki.md

```diff
@@ -0,0 +1,169 @@
+
+The Algorithmic Transaction MUST NOT generate a transaction reciept with a `TransactionType` of `ALG_TX_TYPE`, it MUST generate a transaction receipt of the transaction it is wrapping (the tx in the `parent` field). Implementations MUST 
+
+### Algorithm specification
+
+Further algorithms MUST be specified via an additional EIP.
+
+Each type of algorithm MUST specify the `ALG_TYPE`, `MAX_SIZE` and `GAS_PENALTY` fields, the `GAS_PENALTY` field MUST only account for verification costs, not storage nor signing.
+
+New algorithms MUST also specify how to recover & verify a valid address (`bytes20`) from the `signature_info` field inside the transaction, the verification function MUST follow the following function signature:
+
+`def verify(signature_info: bytes, parent_hash: bytes32) -> boolean, bytes20`
+
+The verify function MUST return `(false, 0x0)` if there was an error recovering a valid address from the signature, otherwise the function MUST return `true` and the address of the signer.
+
+### The default secp256k1 `0x0` algorithmic type
+
+| Constant | Value |
+| - | - |
+| `ALG_TYPE` | `Bytes1(0x0)` |
+| `GAS_PENALTY`| `0` |
+| `MAX_SIZE` | `65` |
+
+```python
+def verify(signature_info: bytes, parent_hash: bytes32) -> boolean, bytes20:
+  assert(len(signature_info) == 65)
+  v, r, s = int.from_bytes(signature_info[0:32], 16), int.from_bytes(signature_info[32:64], 16), signature_info[64]
+
+  # This assumes `ecrecover` is identical to the `ecrecover` function in solidity.
+  signer = ecrecover(parent_hash, v, r, s)
+
+  return (signer != 0x0, signer[:20])
+```
+
+### Verification
+
+Implementations MUST NOT gossip transactions that have a `len(tx.signature_info)` of more than `alg.MAX_SIZE`.
+
+The validity/processing of the transaction is completed by the function defined below:
+```python
+
+def process_transaction(tx: Transaction, from_address: bytes20 = None, start_gas = 21000):
+  match tx:
+    # Verification for other transactions, if `from_address != None`
+    # then the verifier MUST NOT attempt to validate the `y_parity`, `r`, `s`
+    # parameters. Additionally, the verification function MUST start the initial gas value
+    # at `start_gas`.
+    ...
+
+    ALG_TX:
+      assert(from_address == None)          # Ensure no double-wrapping
+      assert(Algorithms[alg_type] != None)  # `Algorithms` is a dictionary containing every defined algorithm
+
+      alg = Algorithms[alg_type]
+
+      assert(len(tx.signature_info) <= alg.MAX_SIZE)
+
+      valid, from_address = alg.verify(tx.signature_info, calculate_signing_hash(wrapped)) # calculate_signing_hash is defined within the wrapped transaction's EIP.
+      assert(valid)
+
+      process_transaction(tx.parent, from_address, 21000 + calculate_penalty(tx.signature_info, alg))
+
+```
+
+### Gas calculation
+
+All transactions that use more resources than the secp256k1 curve suffer an additional penalty. This penalty MUST be calculated as follows:
```

Comment:
I would not specify this struct rule in this EIP, but rather make this "gas penalty" reflect the costs of the recovery algorithm. For instance if something is 10x more CPU intensive than `secp256k1` then that should definitely be included in the gas costs.

Reply:
Oh, nvm, that's already in there :smile: